packages/pigeon/lib/linux_generator.dart

```diff
@@ -0,0 +1,1309 @@
+        args.add(_fromFlValue(generatorOptions.module, field.type,
+            'fl_value_get_list_value(values, $i)'));
+      }
+      indent.writeln('return ${methodPrefix}_new(${args.join(', ')});');
+    });
+  }
+
+  @override
+  void writeFlutterApi(
+    LinuxOptions generatorOptions,
+    Root root,
+    Indent indent,
+    Api api, {
+    required String dartPackageName,
+  }) {
+    final String className = _getClassName(generatorOptions.module, api.name);
+
+    final String methodPrefix =
+        _getMethodPrefix(generatorOptions.module, api.name);
+
+    indent.newln();
+    _writeObjectStruct(indent, generatorOptions.module, api.name, () {
+      indent.writeln('FlBinaryMessenger* messenger;');
+    });
+
+    indent.newln();
+    _writeDefineType(indent, generatorOptions.module, api.name);
+
+    indent.newln();
+    _writeDispose(indent, generatorOptions.module, api.name, () {
+      _writeCastSelf(indent, generatorOptions.module, api.name, 'object');
+      indent.writeln('g_clear_object(&self->messenger);');
+    });
+
+    indent.newln();
+    _writeInit(indent, generatorOptions.module, api.name, () {});
+
+    indent.newln();
+    _writeClassInit(indent, generatorOptions.module, api.name, () {});
+
+    indent.newln();
+    indent.addScoped(
+        '$className* ${methodPrefix}_new(FlBinaryMessenger* messenger) {', '}',
+        () {
+      _writeObjectNew(indent, generatorOptions.module, api.name);
+      indent.writeln('self->messenger = g_object_ref(messenger);');
+      indent.writeln('return self;');
+    });
+
+    for (final Method method in api.methods) {
+      final String methodName = _snakeCaseFromCamelCase(method.name);
+
+      final List<String> asyncArgs = <String>['$className* object'];
+      for (final Parameter param in method.parameters) {
+        final String paramName = _snakeCaseFromCamelCase(param.name);
+        final String paramType = _getType(generatorOptions.module, param.type);
+        asyncArgs.add('$paramType $paramName');
+      }
+      asyncArgs.addAll(<String>[
+        'GCancellable* cancellable',
+        'GAsyncReadyCallback callback',
+        'gpointer user_data'
+      ]);
+      indent.newln();
+      indent.addScoped(
+          "void ${methodPrefix}_${methodName}_async(${asyncArgs.join(', ')}) {",
+          '}',
+          () {});
+
+      final List<String> finishArgs = <String>[
+        '$className* object',
+        'GAsyncResult* result'
+      ];
+      final String returnType =
+          _getType(generatorOptions.module, method.returnType, isOutput: true);
+      if (returnType != 'void') {
+        finishArgs.add('$returnType* return_value');
+      }
+      finishArgs.add('GError** error');
+      indent.newln();
+      indent.addScoped(
+          "gboolean ${methodPrefix}_${methodName}_finish(${finishArgs.join(', ')}) {",
+          '}', () {
+        indent.writeln('return TRUE;');
+      });
+    }
+  }
+
+  @override
+  void writeHostApi(
```

Comment:
This method is huge; can we extract some well-named helpers for specific sections?